nbclient/client.py

```diff
@@ -308,7 +316,49 @@ def start_kernel_manager(self):
         self.km.client_class = 'jupyter_client.asynchronous.AsyncKernelClient'
         return self.km
 
-    async def start_new_kernel_client(self, **kwargs):
+    def _cleanup_kernel(self):
+        try:
+            # Send a polite shutdown request
+            self.kc.shutdown()
+            loop = get_loop()
+            try:
+                # Queue the manager to kill the process, sometimes the built-in and above
+                # shutdowns have not been successful or called yet, so give a direct kill
+                # call here and recover gracefully if it's already dead.
+                try:
+                    # For AsyncKernelManager
+                    loop.run_until_complete(self.km.shutdown_kernel(now=True))
+                except TypeError:
+                    self.km.shutdown_kernel(now=True)
```

Comment:
Should we check if `self.km.shutdown_kernel` is a coroutine, instead of using `try/except`?
Also, we might want to have an async version of `_cleanup_kernel`, which is use in blocking and async functions.

Reply:
I can change to that pattern if you prefer.